### PR TITLE
Reservation handling

### PR DESCRIPTION
**Pull request by sbckr**
covers #10

### CODE REVIEW

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java:207]: Would rather use `fragmentNew` and `fragmentRest` here.

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:212]: Used on only once in remainder of body -> redundant

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:273]: Used on only once in remainder of body -> redundant

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:302]: Used on only once in remainder of body -> redundant

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:381]: Used on only once in remainder of body -> redundant

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/UnlockReservationSupplierTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/UnlockReservationSupplierTest.java:5]: Superfluous space

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntityTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntityTest.java:2]: New file -> 2022

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java:2]: New file -> 2022

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java:354]: Add empty line

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceIT.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceIT.java:2]: New file -> 2022

File: castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java:2]: New file -> 2022

File: castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java:3]: amphora -> castor (check also for other files)

File: castor-service/src/test/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketServiceTest.java
Comment (strieflin) [inline castor-service/src/test/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketServiceTest.java:160]: Misnomer? TupleChunk is not empty!

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:212]: After looking through the whole PR, I see that this is a pattern. Set all values relevant at the beginning then use them throughout the test. I see the rationale behind it. I leave it up to you to decide whether to keep this style or not. I think it helps by adding structure on the one hand. On the other hand it generates warnings in the IDE.

**Review by strieflin — CHANGES_REQUESTED** [review]

File: castor-service/src/test/java/io/carbynestack/castor/service/download/DefaultTuplesDownloadServiceAsMasterIT.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/download/DefaultTuplesDownloadServiceAsMasterIT.java:103]: `setUp` does access the Cache for reservations and telemetry data. In order to access both it requires to have the Spring application context and the autowired `CacheManager` to be available. However `BeforeClass` is executed in a static context making both not available.

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java]: No, just a typo ;-) -> fixed

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java:73]: fixed by changing reference from field to accessor

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java]: fixed

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java]: updated log message

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageService.java]: fixed

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntity.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntity.java]: replaced by lombok-generated equals and hashCode

File: castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntity.java
Comment (sbckr) [reply castor-service/src/main/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntity.java]: fixed

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/CreateReservationSupplierTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/CreateReservationSupplierTest.java:2]: this file was actually moved from package `io.carbynestack.castor.service.download` to `io.carbynestack.castor.service.persistence.cache` and updated afterwards. However this is not recognized by git for some reasons

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java:108]: see [above](https://github.com/carbynestack/castor/pull/29#discussion_r869349328)
null check to behave like a singleton

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java:138]: one could argue how to verify a returned object remains untouched if it is not verified that is was returned before at all?

code could also be replaced by sth like the following
```java
int consecutiveCalls = 3;
for(int i = 0; i < consecutiveCalls; i++) {
  assertEquals(
        testReservation,
        reservationCachingService.getUnlockedReservation(
            testReservation.getReservationId(), testTupleType, testNumberReservedTuples));
}
```
each call should return the exact same result

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:212]: Did it for the structure but have observed the warnings too. Happy to discuss which way to go

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java:207]: I think this would reflect the intention as the purpose of the test is to verify that 
- the `applyReservation` leaves fragments that reference sections _before_ the actual tuples of interest untouched
- applies the reservation on multiple fragments if the tuples of interest are actually described by / split to multiple consecutive fragments

example:
[Fragment FA withs Tuple T0-T3] [Fragment FB with Tuples T4-T5] [Fragment FC with Tuple T6]
The reservation is to be applied on tuples T4-T6
expected result:
- First fragment FA, which describes a section before the first tuple of interest, remains untouched
- Reservation is applied on Fragment FB and FC

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:274]: see discussion [above](https://github.com/carbynestack/castor/pull/29#discussion_r868054674)

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:303]: see discussion [above](https://github.com/carbynestack/castor/pull/29#discussion_r868054674)

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:382]: see discussion [above](https://github.com/carbynestack/castor/pull/29#discussion_r868054674)

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/UnlockReservationSupplierTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/UnlockReservationSupplierTest.java]: removed

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntityTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentEntityTest.java]: fixed

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java]: fixed

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceTest.java:354]: hmmm... supposed to be there

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceIT.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/fragmentstore/TupleChunkFragmentStorageServiceIT.java]: fixed

File: castor-service/src/test/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketServiceTest.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketServiceTest.java]: renamed

File: castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java:3]: fixed

File: castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java
Comment (sbckr) [reply castor-service/src/test/java/io/carbynestack/castor/service/util/TupleChunkFragmentEntityListMatcher.java:2]: fixed

File: castor-service/src/main/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketService.java
Comment (strieflin) [reply castor-service/src/main/java/io/carbynestack/castor/service/websocket/DefaultCastorWebSocketService.java:98]: I see. That makes sense. Doing the casting probably creates more noise. Keep it as it is.

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java
Comment (strieflin) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceIT.java:138]: I see. Ok, that's fine for me then.

File: castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java
Comment (strieflin) [reply castor-service/src/test/java/io/carbynestack/castor/service/persistence/cache/ReservationCachingServiceTest.java:212]: Let's stick with the way you did it and ignore the warnings.

**Review by strieflin — APPROVED** [review]
LGTM

**Comment by strieflin** [comment]
Cannot merge this @sbckr as not all of your commits are signed.

**Comment by sbckr** [comment]
> Cannot merge this @sbckr as not all of your commits are signed.

@strieflin 
Thank you for your review. All commits are now signed accordingly

**Review by strieflin — APPROVED** [review]
LGTM

**Pull request merged**
